### Tarefa do Encontro 1 - Métodos de Classificação Edit KNN

### Aluno: Henrique Seschin Neto: R.A: 202083

### Aluno: Robson Agapito Correa: R.A: 202062


## 01 - K-Nearest Neighbors

---

Vamos iniciar carregando os dados, embaralhando as amostras, e separando 100 amostras para teste e 50 amostras para treino,

In [235]:
import numpy as np
import pandas as pd

df_dataset = pd.read_csv('iris.csv', sep=',', index_col=None)

In [236]:
df_dataset.head()

comprimento_sepala  largura_sepala  ...  largura_petala       classe
0                 5.1             3.5  ...             0.2  Iris-setosa
1                 4.9             3.0  ...             0.2  Iris-setosa
2                 4.7             3.2  ...             0.2  Iris-setosa
3                 4.6             3.1  ...             0.2  Iris-setosa
4                 5.0             3.6  ...             0.2  Iris-setosa

[5 rows x 5 columns]

In [237]:
from sklearn.utils import shuffle
df_dataset = shuffle(df_dataset, random_state=42)

In [238]:
df_dataset.head()

comprimento_sepala  largura_sepala  ...  largura_petala           classe
73                  6.1             2.8  ...             1.2  Iris-versicolor
18                  5.7             3.8  ...             0.3      Iris-setosa
118                 7.7             2.6  ...             2.3   Iris-virginica
78                  6.0             2.9  ...             1.5  Iris-versicolor
76                  6.8             2.8  ...             1.4  Iris-versicolor

[5 rows x 5 columns]

In [239]:
treino = df_dataset[:100].values
teste = df_dataset[100:].values

print(treino.shape)
print(teste.shape)

(100, 5)
(50, 5)


---

Em seguida, vamos dar início à implementação do algoritmo KNN. O algoritmo funciona basicamente nas seguintes etapas:

- Calcular a distância de uma amostra para todas as demais;
- Pegar as n amostras mais próximas;
- Computar qual a classe mais frequente.

### Calculando a distância

In [240]:
#  testes explicando:

In [241]:
teste[0]

array([4.6, 3.1, 1.5, 0.2, 'Iris-setosa'], dtype=object)

In [242]:
treino[0]

array([6.1, 2.8, 4.7, 1.2, 'Iris-versicolor'], dtype=object)

In [243]:
#[0,:4 slice para não pegar o nome das colunas

In [244]:
# diferença de uma amostra de teste, para cada amostra da base de treino.

amostra_teste = teste[0,:4]
amostra_treino = treino[0,:4]

distancia = np.sum((amostra_treino - amostra_teste) ** 2) ** 0.5
print("Distância entre duas amostras quaisquer:", distancia)

Distância entre duas amostras quaisquer: 3.685105154537656


In [245]:
amostra_treino - amostra_teste

array([1.5, -0.30000000000000027, 3.2, 1.0], dtype=object)

Também é possível calcular a distância para todas as amostras de uma só vez.

In [246]:
# axis=1  - pega apenas as linhas na somatória

distancias = np.sum((amostra_teste - treino[:,:4]) ** 2, axis=1) ** 0.5
print("Distâncias de uma amostra de teste para todas as amostras de treino:")
print(distancias)

Distâncias de uma amostra de teste para todas as amostras de treino:
[3.685105154537656 1.3228756555322956 6.590144156238162 3.5623026261113755
 4.154515615568198 0.8774964387392128 2.5806975801127883 4.760252094164762
 3.749666651850535 2.891366458960192 4.451965857910413 0.26457513110645925
 1.0049875621120892 0.3162277660168386 0.8660254037844384
 3.889730067755345 5.109794516416487 2.816025568065745 3.3926390907374753
 4.913247398615299 0.17320508075688812 4.047221268969612
 0.5477225575051663 4.873397172404482 6.215303693304134 4.7455242070818695
 5.081338406365 5.349766350038103 0.26457513110645925 0.22360679774997916
 0.7071067811865476 1.714642819948225 3.776241517699842 0.3741657386773941
 0.29999999999999954 4.288356328478313 3.733630940518894
 0.7211102550927982 0.6480740698407862 1.1704699910719625
 4.177319714841084 3.606937759374287 4.042276586281548 1.1661903789690604
 1.0000000000000007 2.6038433132583076 4.198809355043403 4.669047011971501
 3.722902093797257 5.78446194

---

Em seguida, escolhemos quantos vizinhos vamos considerar e escolhemos os mais próximos da amostra avaliada.
A partir dos índices encontrados, calculamos qual é a classe mais frequente e atribuímos à amostra.

In [247]:
# argsort - pega os valores menores e ordena eles na posição, mantendo seu indice original
# k = 5, são os registros das menores 5 distancias 
k = 5
indices = distancias.argsort()[:k]
print("Indices dos vizinhos mais próximos:", indices)

Indices dos vizinhos mais próximos: [68 20 29 11 28]


In [248]:
from collections import Counter

#pede a coluna 4 na base original
vizinhos = treino[indices,4]
print("Vizinhos mais próximos:", vizinhos)

# calculando a moda do numero de K vizinhos, depois veriificar o mais comun, 
#para tirar da tupla, indice [0]
classe = Counter(vizinhos).most_common(1)[0][0]
print("Classe mais frequente e quantidade:", Counter(vizinhos).most_common(1))

Vizinhos mais próximos: ['Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa']
Classe mais frequente e quantidade: [('Iris-setosa', 5)]


---

Para verificar se a classe encontrada pelo modelo está correta, basta comparar a saída com o rótulo da amostra.

In [249]:
# verificar se a aclsse esta certa, compara a base de teste original com a classficiada

In [250]:
classe_certa = teste[0, 4]
print(classe_certa)
if classe_certa == classe:
    print("Correto")
else:
    print("Errado")

Iris-setosa
Correto


---

## Exercícios

### (1) Estruture o código para funcionar como uma função, que tem a seguinte assinatura:

**knn(X, y, xt)**

Onde X representa as amostras de treino, y representa as classes do treino e xt representa uma amostra qualquer a ser rotulada.

#### Função Knn

In [251]:
# implemente seu código aqui

def knn(X,y,xt):
    
    #calcula as distancia 
    distancias = np.sum((xt - X) ** 2, axis =1) ** 0.5
       
    #ordendar
    indices = distancias.argsort()
      
    #selecionar as amostras baseadas no k=1
    indice = indices[0]
      
    #calcular a moda, encontrar a classe baseada no indice, porque k=1
    classe = y[indice]
    
    
    return classe

In [252]:
# teste

X = treino[:,:4]
y = treino[:,4]
xt = teste[1,:4]

knn(X,y,xt)

'Iris-setosa'

In [253]:
yt = teste[1,4]
yt

'Iris-setosa'

### (2) A partir da função anterior implementada, utilize-a para gerar as variações Edit KNN por remoção e inserção. Verifique o resultado final observando: a quantidade de amostras e a acurácia do conjunto de teste.

#### Função Edit KNN Inserção

In [254]:
# implemente seu código aqui

# Edit KNN para a função de inserção

def editKNN_insercao(X,y):
    
    #iniciar uma base reduzida
    X_reduzido,y_reduzido = [],[]
    
    #para cada amostra da base original:
    for X_atual, y_atual in zip(X,y):
        
        if len(X_reduzido) == 0:
            X_reduzido.append(X_atual)
            y_reduzido.append(y_atual)
            continue    
        
    #verifica se a base reduzida classifica corretamente
        if knn(X_reduzido,y_reduzido,X_atual) != y_atual:
            X_reduzido.append(X_atual)
            y_reduzido.append(y_atual)
     
    return X_reduzido,y_reduzido

editKNN_insercao(X,y)

([array([6.1, 2.8, 4.7, 1.2], dtype=object),
  array([5.7, 3.8, 1.7, 0.3], dtype=object),
  array([7.7, 2.6, 6.9, 2.3], dtype=object),
  array([6.9, 3.1, 5.1, 2.3], dtype=object),
  array([6.1, 3.0, 4.9, 1.8], dtype=object),
  array([6.0, 3.4, 4.5, 1.6], dtype=object),
  array([6.7, 3.1, 4.7, 1.5], dtype=object),
  array([5.9, 3.2, 4.8, 1.8], dtype=object),
  array([6.0, 2.7, 5.1, 1.6], dtype=object),
  array([6.1, 2.6, 5.6, 1.4], dtype=object),
  array([6.0, 2.2, 5.0, 1.5], dtype=object)],
 ['Iris-versicolor',
  'Iris-setosa',
  'Iris-virginica',
  'Iris-virginica',
  'Iris-virginica',
  'Iris-versicolor',
  'Iris-versicolor',
  'Iris-versicolor',
  'Iris-versicolor',
  'Iris-virginica',
  'Iris-virginica'])

#### Aplicando a função inserção no knn com a base de treino reduzida

In [255]:
# aplicando a função inserção no knn com a base de treino reduzida

X_reduz, Y_reduz =  editKNN_insercao(X,y)

X_teste = teste[:,:4]
y_teste = teste[:,4]

classe = []

for amostra in X_teste:
    classe.append(knn(X_reduz, Y_reduz, amostra))
acuracia_tit = (y_teste == classe).mean()

print('A acurácia usando a função de inserção na base de treino reduzida foi de: %f' % acuracia_tit)   
    

A acurácia usando a função de inserção na base de treino reduzida foi de: 0.980000


In [256]:
#tamanho da base X reduzida pela função Inserção 
len(X_reduz)

11

In [257]:
#tamanho da base Y reduzida pela função Inserção 
len(Y_reduz)

11

####  Função Edit KNN  Remoção

In [258]:
# implemente seu código aqui

# Edit KNN para a função de remoção

def editKNN_remocao(X,y):
    
    #iniciar uma base reduzida:
    X_reduzido,y_reduzido = [],[]
    pos = 0
    
    #para cada amostra da base original:
    for X_atual, y_atual in zip(X,y):
        #copiar base padrão
        X_novo = X.copy()
        y_novo = y.copy()
        
        #deleta os registros que ira comparar:
        X_novo = np.delete(X_novo, pos, 0)
        y_novo = np.delete(y_novo, pos, 0)
        pos = pos + 1   
            
        #verifica se a base reduzida classifica corretamente
        res = knn(X_novo,y_novo,X_atual)
        
        if res != y_atual:
            X_reduzido.append(X_atual)
            y_reduzido.append(y_atual)  
        
    return X_reduzido,y_reduzido

editKNN_remocao(X,y)    

([array([6.2, 2.2, 4.5, 1.5], dtype=object),
  array([6.5, 3.2, 5.1, 2.0], dtype=object),
  array([6.3, 2.8, 5.1, 1.5], dtype=object),
  array([5.9, 3.2, 4.8, 1.8], dtype=object),
  array([6.0, 2.7, 5.1, 1.6], dtype=object),
  array([6.1, 2.6, 5.6, 1.4], dtype=object),
  array([6.0, 2.2, 5.0, 1.5], dtype=object)],
 ['Iris-versicolor',
  'Iris-virginica',
  'Iris-virginica',
  'Iris-versicolor',
  'Iris-versicolor',
  'Iris-virginica',
  'Iris-virginica'])

#### Aplicando a função remoção no knn com a base de treino reduzida

In [259]:
# aplicando a função remoção no knn com a base de treino reduzida

X_reduz, Y_reduz =  editKNN_remocao(X,y)

X_teste = teste[:,:4]
y_teste = teste[:,4]

classe = []

for amostra in X_teste:
    classe.append(knn(X_reduz, Y_reduz, amostra))
acuracia_tit = (y_teste == classe).mean()

print('A acurácia usando a função de remoção na base de treino reduzida foi de: %f' % acuracia_tit)   

A acurácia usando a função de remoção na base de treino reduzida foi de: 0.560000


In [260]:
#tamanho da base X reduzida pela função Remoção
len(X_reduz)

7

In [261]:
#tamanho da base X reduzida pela função Remoção
len(Y_reduz)

7

### Respostas:

#### A acurácia no EditKnn da função inserção, para o conjunto de teste, foi de 98%
#### A acurácia no EditKnn da função remoção, para o conjunto de teste,  foi de 56%

#### A quantidade de amostras na base de teste reduzida no EditKnn da função inserção passou a ser 11
#### A quantidade de amostras na base de teste reduzida no EditKnn da função remoção passou a ser 7

#### Em ambas as funções, a base de treino foi reduzida.
#### A função Edit Knn Inserção, obteve a melhor acurácia, utilizando da base de treino reduzida.

### (3) Aplique o algoritmo Edit KNN na base do Titanic e verifique o resultado final e a quantidade de amostras mantidas.

### O arquivo **titanic_data.csv** contém informações reais sobre os passageiros organizadas da seguinte forma:

- Survived (Sobreviveu): 0 = Não, 1 = Sim
- Pclass (Classe): Classe de ingresso 1 = 1º, 2 = 2º, 3 = 3º
- Sex (Sexo): Sexo do passageiro
- Age (Idade): Idade em anos
- Sibsp: Quantidade de irmãos / cônjuges a bordo do Titanic
- Parch: Quantidade de pais / crianças a bordo do Titanic
- Ticket (Bilhete): Número do bilhete de embarque
- Fare (Tarifa): Tarifa paga pelo Passageiro
- Cabin (Cabine): Número de cabine
- Embarked (Embarque): Porto de Embarque (C = Cherbourg, Q = Queenstown, S = Southampton)

In [262]:
#Carregamento da base original:
df_dataset_tit = pd.read_csv('titanic_data.csv',sep=',', index_col=None)

In [263]:
df_dataset_tit.head(3)  

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S

[3 rows x 12 columns]

In [264]:
#Resumo das informações:
df_dataset_tit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [265]:
# A coluna "Cabin" tem muitos dados faltantes,porém não são dados relevantes. Excluindo a coluna "Cabin"
df_dataset_tit.drop('Cabin',axis=1,inplace=True)

In [266]:
df_dataset_tit.head(3)  

PassengerId  Survived  Pclass  ...            Ticket     Fare  Embarked
0            1         0       3  ...         A/5 21171   7.2500         S
1            2         1       1  ...          PC 17599  71.2833         C
2            3         1       3  ...  STON/O2. 3101282   7.9250         S

[3 rows x 11 columns]

In [267]:
#achando a idade média entre sexo - "Sex" e classes - "Pclass".
idade_media = round(df_dataset_tit.groupby(['Pclass','Sex'])['Age'].mean().unstack(),0)

print('Média de idade da primeira classe é',round(idade_media.loc[1].mean()))
print('Média de idade da segunda classe é' ,round(idade_media.loc[2].mean()))
print('Média de idade da terceira classe é',round(idade_media.loc[3].mean()))

Média de idade da primeira classe é 38
Média de idade da segunda classe é 30
Média de idade da terceira classe é 24


In [268]:
# Função para completar a idade média dos passageiros nos dados faltantes da coluna idade - "Age", 
# levando em consideração o sexo - "Sex" e a classe de embarque do novaio - "Pclass"

def muda_coluna_idade(coluna):
    age = coluna['Age']
    pclass = coluna['Pclass']
    sex = coluna['Sex']  
        
    if pd.isnull(age):
        return idade_media.loc[pclass][sex]
    else:
        return age

In [269]:
#inserindo a idade média nos dados faltantes da coluna idade - "Age"
df_dataset_tit['Age'] = df_dataset_tit[['Age', 'Pclass','Sex']].apply(muda_coluna_idade, axis = 1)

In [270]:
#Removendo valores de colunas não utilizadas
df_dataset_tit.drop(['PassengerId','Embarked','Name','Ticket','SibSp','Parch'],axis=1,inplace=True)

In [271]:
#visualizando a base 
df_dataset_tit.head(5)  

Survived  Pclass     Sex   Age     Fare
0         0       3    male  22.0   7.2500
1         1       1  female  38.0  71.2833
2         1       3  female  26.0   7.9250
3         1       1  female  35.0  53.1000
4         0       3    male  35.0   8.0500

In [272]:
#resumo das informações - Base Finalizada.
df_dataset_tit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    object 
 3   Age       891 non-null    float64
 4   Fare      891 non-null    float64
dtypes: float64(2), int64(2), object(1)
memory usage: 34.9+ KB


In [273]:
#Misturando os dados da base:
df_dataset_tit = shuffle(df_dataset_tit, random_state=42)

In [274]:
#Reordenandos as colunas:
df_dataset_tit = df_dataset_tit[['Survived','Pclass','Age','Fare','Sex']]

In [275]:
df_dataset_tit.head(5)  

Survived  Pclass   Age     Fare     Sex
709         1       3  27.0  15.2458    male
439         0       2  31.0  10.5000    male
840         0       3  20.0   7.9250    male
720         1       2   6.0  33.0000  female
39          1       3  14.0  11.2417  female

In [276]:
df_dataset_tit['Sex'] = df_dataset_tit['Sex'].apply(lambda x:1 if x=='male' else 0)

In [277]:
df_dataset_tit.head(5) 


Survived  Pclass   Age     Fare  Sex
709         1       3  27.0  15.2458    1
439         0       2  31.0  10.5000    1
840         0       3  20.0   7.9250    1
720         1       2   6.0  33.0000    0
39          1       3  14.0  11.2417    0

In [278]:
#normalizando os dados:
#passando o standscaler nas colunas de X ["Pclass, Age e Fare"]

from sklearn.preprocessing import StandardScaler

tit_normalizado = df_dataset_tit.copy()

col_names = ['Pclass','Age','Fare','Sex']
features = tit_normalizado[col_names]
scaler = StandardScaler().fit(features.values)
features = scaler.transform(features.values)

#### Base do Titanic Normalizada

In [279]:
#base titanic normalizada
tit_normalizado[col_names] = features
tit_normalizado.head()

Survived    Pclass       Age      Fare       Sex
709         1  0.827377 -0.179804 -0.341452  0.737695
439         0 -0.369365  0.122019 -0.437007  0.737695
840         0  0.827377 -0.707993 -0.488854  0.737695
720         1 -0.369365 -1.764371  0.016023 -1.355574
39          1  0.827377 -1.160726 -0.422074 -1.355574

#### Separando a base do Titanic em Treino (70%) e Teste (30%):

In [280]:
# Total da base = 891 amostras:
# 70% para Treino = (624 amostras)
# 30% para Teste  = (267 amostras)

df_titanic = tit_normalizado

treino_titanic = df_titanic[:624].values
teste_titanic = df_titanic[624:].values

print(treino_titanic.shape)
print(teste_titanic.shape)

(624, 5)
(267, 5)


In [281]:
# diferença de uma amostra de teste, para uma determinada amostra da base de treino do Titanic.
amostra_teste_tit = teste_titanic[0,1:5] 
amostra_treino_tit = treino_titanic[0,1:5]

distancia = np.sum((amostra_treino_tit - amostra_teste_tit) ** 2) ** 0.5
print("Distância entre duas amostras quaisquer:", distancia)

Distância entre duas amostras quaisquer: 3.311920174497596


In [282]:
amostra_treino_tit - amostra_teste_tit

array([ 2.39348417, -1.88638977, -1.29675842,  0.        ])

In [283]:
# diferença de uma amostra de teste, para cada amostra da base de treino do Titanic.
# axis=1  - pega apenas as linhas na somatória

distancias = np.sum((amostra_teste_tit - treino_titanic[:,:4]) ** 2, axis=1) ** 0.5
print("Distâncias de uma amostra de teste para todas as amostras de treino do Titanic:")
print(distancias)

Distâncias de uma amostra de teste para todas as amostras de treino do Titanic:
[3.13224303 2.97263477 2.73800767 4.33727242 3.63050547 4.33606928
 3.34040556 2.85225299 3.56829011 4.58835352 3.6617617  2.17847447
 2.5051469  2.36446819 2.86956557 4.63773115 4.17168903 3.33981867
 3.05203804 3.66307055 2.56616128 4.28822797 3.44805783 2.72638241
 2.38984685 3.12270542 3.72458239 3.58511419 3.48504588 2.32980462
 2.78475    3.34040556 3.70935782 2.65254944 3.48533028 2.39026919
 4.34384425 3.34040556 4.32455701 2.45357688 3.04832825 2.39635321
 2.45202552 2.45504643 2.11534442 2.75118498 2.78242498 2.59448872
 2.57434842 3.92596992 4.02087019 4.25515394 1.98218077 4.88295536
 2.78196057 4.54248144 2.95237169 5.53162362 3.47920347 3.4478786
 2.69421364 4.08461998 3.572317   3.7304318  2.45504643 3.06852284
 3.04832825 2.38772699 2.88713052 4.23034922 2.9953024  5.98090491
 4.40142435 4.17507463 2.60737333 2.14126619 2.34466335 4.44234313
 2.84948186 2.04710078 2.72616857 3.6391002  4.366

#### Utilizando K=3, para a base do Titanic

In [284]:
#Ajustando o valor de K para a base do Titanic
# k = 3, são os registros das menores 3 distancias 
k = 3
indices = distancias.argsort()[:k]
print("Indices dos vizinhos mais próximos:", indices)

Indices dos vizinhos mais próximos: [354 550 119]


In [285]:
from collections import Counter

#pega a coluna 0 na base original do Titanic - Coluna - Survived (0 ou 1)
vizinhos = treino_titanic[indices,0]
print("Vizinhos mais próximos:", vizinhos)

# calculando a moda do numero de K vizinhos, depois verificar o mais comun,para tirar da tupla, indice [0]
classe = Counter(vizinhos).most_common(1)[0][0]
print("Classe mais frequente e quantidade:", Counter(vizinhos).most_common(1))

Vizinhos mais próximos: [0. 0. 0.]
Classe mais frequente e quantidade: [(0.0, 3)]


#### Definindo a base X e a saída Y da base do Iitanic

In [286]:
# definindo a base X e a saída Y da base do titanic

X_tit = treino_titanic[:,1:5]   # X = todo o resto da base, menos a coluna "Survived"
y_tit = treino_titanic[:,0]     # y é a saída, a resposta, apenas a coluna 0, onde 0: não sobreviveu, 1: sobreviveu

xt_tit = teste_titanic[0,1:5]   # um teste do Knn - uma simples amostra para ser titulada

knn(X_tit,y_tit,xt_tit)

0.0

#### Aplicando a função de Edit Knn Inserção na base do Titanic

#### Aplicando a função Edit Knn inserção no knn com a base de treino reduzida, no conjunto de teste.

In [287]:
# Aplicando a função inserção no knn com a base de treino reduzida, no conjunto de teste

Xtit_reduzi_inser, Ytit_reduzi_inser =  editKNN_insercao(X_tit,y_tit)

X_teste_tit = teste_titanic[:,1:5]
y_teste_tit = teste_titanic[:,0]

classe_inser = []

for amostra_inser in X_teste_tit:
    classe_inser.append(knn(Xtit_reduzi_inser, Ytit_reduzi_inser, amostra_inser))
acuracia_tit = (y_teste_tit == classe_inser).mean()

print('A acurácia para o Knn, aplicando a função de inserção e usando a base de treino reduzida foi de: %f' % acuracia_tit)   
    

A acurácia para o Knn, aplicando a função de inserção e usando a base de treino reduzida foi de: 0.692884


In [288]:
#tamanho da base X reduzida pela função Inserção
len(Xtit_reduzi_inser)

187

In [289]:
#tamanho da base Y reduzida pela função Inserção
len(Ytit_reduzi_inser)

187

#### Aplicando a função Edit Knn Remoção na base do Titanic

#### Aplicando a função Edit Knn Remoção no knn com a base de treino reduzida, no cojunto de teste

In [290]:
# Aplicando a função remoção no knn com a base de treino reduzida

Xtit_reduzi_remocao, Ytit_reduzi_remocao =  editKNN_remocao(X_tit,y_tit)

X_teste_tit = teste_titanic[:,1:5]
y_teste_tit = teste_titanic[:,0]

classe_rem = []

for amostra_rem in X_teste_tit:
    classe_rem.append(knn(Xtit_reduzi_remocao, Ytit_reduzi_remocao, amostra_rem))
acuracia_tit = (y_teste_tit == classe_rem).mean()

print('A acurácia para o Knn, aplicando a função de remoção e usando a base de treino reduzida foi de: %f' % acuracia_tit)   

A acurácia para o Knn, aplicando a função de remoção e usando a base de treino reduzida foi de: 0.475655


In [291]:
#tamanho da base X reduzida pela função Remoção
len (Xtit_reduzi_remocao)

145

In [292]:
#tamanho da base Y reduzida pela função Remoção
len(Ytit_reduzi_remocao)

145

### Respostas:

#### A acurácia foi de 69% na base reduzida pelo método da inserção e 48% na base reduzida pelo método da remoção, para o conjunto de teste.

#### A quantidade de amostras, ficou um pouco menor na base reduzida da função Edit Knn Remoção.    

#### A função Edit Knn Remoção ficou com uma base reduzida de 145 amostras e a função Edit Knn Inserção ficou com uma base reduzida de 187 amostras.

#### Usando o Knn com as funções Edit, teve uma melhora na redução da quantidade de amostras na base de treino de ambas as funções. 